File: ConnorPet/Sources/ConnorPet/SpriteSheet.swift
import AppKit
import CoreGraphics
import Foundation

/// One playable animation: its frames (already-cropped images) and the
/// per-frame hold durations to play them at (mirrors Orca's SpriteAnimation).
struct SpriteAnimationFrames {
    let images: [NSImage]
    let durationsMs: [Double]
}

/// Loads `spritesheet.png` + `pet.json` and pre-slices every declared
/// animation into individual frame images, so the pet view just indexes an
/// array on each timer tick instead of doing CSS-style background-position math.
final class SpriteSheet {
    let manifest: PetManifest
    private var animationCache: [String: SpriteAnimationFrames] = [:]
    private let cgImage: CGImage

    init(manifestURL: URL, spritesheetURL: URL) throws {
        manifest = try PetManifest.load(from: manifestURL)
        let data = try Data(contentsOf: spritesheetURL)
        guard let source = CGImageSourceCreateWithData(data as CFData, nil),
              let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
            throw NSError(domain: "SpriteSheet", code: 1, userInfo: [NSLocalizedDescriptionKey: "Could not decode spritesheet.png"])
        }
        cgImage = image

        let expectedW = manifest.frame.width * (image.width / manifest.frame.width)
        let expectedH = manifest.frame.height * (image.height / manifest.frame.height)
        guard image.width == expectedW, image.height == expectedH else {
            throw NSError(domain: "SpriteSheet", code: 2, userInfo: [
                NSLocalizedDescriptionKey: "spritesheet \(image.width)x\(image.height) is not a clean multiple of frame \(manifest.frame.width)x\(manifest.frame.height)"
            ])
        }
    }

    /// Loads a bundled pet by slug from `Resources/pets/<slug>/`. Shared by the
    /// menu-bar picker and the battle screen so both resolve characters the same way.
    static func bundled(slug: String) throws -> SpriteSheet {
        guard
            let spritesheetURL = AppDelegate.resourceBundle.url(forResource: "spritesheet", withExtension: "png", subdirectory: "pets/\(slug)"),
            let manifestURL = AppDelegate.resourceBundle.url(forResource: "pet", withExtension: "json", subdirectory: "pets/\(slug)")
        else {
            throw NSError(domain: "ConnorPet", code: 1, userInfo: [NSLocalizedDescriptionKey: "missing bundled resources for pet '\(slug)'"])
        }
        return try SpriteSheet(manifestURL: manifestURL, spritesheetURL: spritesheetURL)
    }

    func animation(named name: String) -> SpriteAnimationFrames? {
        if let cached = animationCache[name] {
            return cached
        }
        guard let anim = manifest.animations[name] else { return nil }
        let frameW = manifest.frame.width
        let frameH = manifest.frame.height

        var images: [NSImage] = []
        images.reserveCapacity(anim.frames)
        for col in 0..<anim.frames {
            // Verified empirically: row index maps directly to y = row*frameHeight
            // for a CGImage decoded from this PNG via CGImageSource — see
            // scripts note in README. No vertical flip needed.
            let rect = CGRect(x: col * frameW, y: anim.row * frameH, width: frameW, height: frameH)
            guard let cropped = cgImage.cropping(to: rect) else { continue }
            images.append(NSImage(cgImage: cropped, size: NSSize(width: frameW, height: frameH)))
        }

        let durations = anim.frameDurationsMs ?? Array(repeating: 1000.0 / manifest.fps, count: anim.frames)
        let frames = SpriteAnimationFrames(images: images, durationsMs: durations)
        animationCache[name] = frames
        return frames
    }

    /// Resolves a live `PetAnimationName` to a manifest animation key,
    /// mirroring `SpriteFrame`'s fallback chain in `PetOverlay.tsx`:
    /// exact name -> defaultAnimation -> first declared animation.
    func resolvedAnimation(for name: PetAnimationName) -> SpriteAnimationFrames? {
        if let exact = animation(named: name.rawValue) {
            return exact
        }
        if let def = manifest.defaultAnimation, let fallback = animation(named: def) {
            return fallback
        }
        if let firstKey = manifest.animations.keys.first {
            return animation(named: firstKey)
        }
        return nil
    }
}

File: apps/windows/pet_win/spritesheet.py
"""스프라이트 시트 로더 (SpriteSheet.swift / PetManifest.swift 포팅).

pet.json 매니페스트 + spritesheet.png 를 읽어 애니메이션별 프레임(QPixmap)과
프레임 지속시간(ms)을 뽑는다. 프레임은 `col*width, row*height` 에서 잘라낸다.
"""
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from typing import Optional

from PySide6.QtGui import QPixmap


@dataclass
class Animation:
    name: str
    frames: list          # list[QPixmap]
    durations_ms: list    # list[float]


class SpriteSheet:
    def __init__(self, slug: str, pet_dir: str):
        self.slug = slug
        manifest_path = os.path.join(pet_dir, "pet.json")
        with open(manifest_path, "r", encoding="utf-8") as f:
            self.manifest = json.load(f)

        frame = self.manifest.get("frame", {})
        self.frame_w = int(frame.get("width", 200))
        self.frame_h = int(frame.get("height", 200))
        self.fps = float(self.manifest.get("fps", 6)) or 6.0
        self.display_name = self.manifest.get("displayName", slug)
        self.default_animation = self.manifest.get("defaultAnimation", "idle")

        sheet_name = self.manifest.get("spritesheetPath", "spritesheet.png")
        self._sheet = QPixmap(os.path.join(pet_dir, sheet_name))
        if self._sheet.isNull():
            raise RuntimeError(f"스프라이트 시트를 못 읽음: {slug}/{sheet_name}")

        self._animations = {}
        for name, spec in (self.manifest.get("animations") or {}).items():
            self._animations[name] = self._build_animation(name, spec)

    def _build_animation(self, name, spec) -> Animation:
        row = int(spec.get("row", 0))
        count = int(spec.get("frames", 1))
        durations = spec.get("frameDurationsMs")
        if not isinstance(durations, list) or len(durations) != count:
            durations = [1000.0 / self.fps] * count
        frames = []
        for col in range(count):
            x = col * self.frame_w
            y = row * self.frame_h
            frames.append(self._sheet.copy(x, y, self.frame_w, self.frame_h))
        return Animation(name=name, frames=frames,
                         durations_ms=[float(d) for d in durations])

    def animation(self, name: str) -> Animation:
        """이름으로 애니메이션을 찾되, 없으면 기본 애니메이션으로 폴백."""
        if name in self._animations:
            return self._animations[name]
        if self.default_animation in self._animations:
            return self._animations[self.default_animation]
        # 그래도 없으면 아무거나 첫 번째.
        return next(iter(self._animations.values()))

    def has_animation(self, name: str) -> bool:
        return name in self._animations
